Fix for subspec options validation issue #27715

pdellaert opened the pull request:
##### SUMMARY
Fixes #27715

This change actually looks at the `required_together`, `required_on_of` and `required_if` fields on the subspec itself when validating the subspec. Where the original code looked at the global values, which caused an issue.

##### ISSUE TYPE
 - Bugfix Pull Request

##### COMPONENT NAME
basic.py

##### ANSIBLE VERSION
```
ansible 2.4.0 (fix_subspec_issue_27715 792bccf1d2) last updated 2017/08/04 00:53:29 (GMT +200)
  config file = None
  configured module search path = [u'/Users/pdellaer/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /Users/pdellaer/GitHub/pdellaert/ansible/lib/ansible
  executable location = /Users/pdellaer/GitHub/pdellaert/ansible/bin/ansible
  python version = 2.7.13 (default, Dec 17 2016, 23:03:43) [GCC 4.2.1 Compatible Apple LLVM 8.0.0 (clang-800.0.42.1)]
```

##### ADDITIONAL INFORMATION
Below is an example on how to configure the subspec values for `required_together`, `required_on_of` and `required_if` (based on the `nuage_vspk` module)
```python
    module = AnsibleModule(
        argument_spec=dict(
            auth=dict(
                required=True,
                type='dict',
                options=dict(
                    api_username=dict(required=True, type='str'),
                    api_enterprise=dict(required=True, type='str'),
                    api_url=dict(required=True, type='str'),
                    api_version=dict(required=True, type='str'),
                    api_password=dict(default=None, required=False, type='str', no_log=True),
                    api_certificate=dict(default=None, required=False, type='str', no_log=True),
                    api_key=dict(default=None, required=False, type='str', no_log=True)
                ),
                required_one_of=[
                    ['api_password', 'api_certificate']
                ],
                required_together=[
                    ['api_certificate', 'api_key']
                ]
            ),
            type=dict(required=True, type='str'),
            id=dict(default=None, required=False, type='str'),
            parent_id=dict(default=None, required=False, type='str'),
            parent_type=dict(default=None, required=False, type='str'),
            state=dict(default=None, choices=['present', 'absent'], type='str'),
            command=dict(default=None, choices=SUPPORTED_COMMANDS, type='str'),
            match_filter=dict(default=None, required=False, type='str'),
            properties=dict(default=None, required=False, type='dict'),
            children=dict(default=None, required=False, type='list')
        ),
        mutually_exclusive=[
            ['command', 'state']
        ],
        required_together=[
            ['parent_id', 'parent_type']
        ],
        required_one_of=[
            ['command', 'state']
        ],
        required_if=[
            ['state', 'present', ['id', 'properties', 'match_filter'], True],
            ['state', 'absent', ['id', 'properties', 'match_filter'], True],
            ['command', 'change_password', ['id', 'properties']],
            ['command', 'wait_for_job', ['id']]
        ],
        supports_check_mode=True
    )
```

pdellaert committed:
Fix for issue ansible/ansible#27715

nitzmahone reviewed (approved):
LGTM- if shippable stays green, good to merge

pdellaert commented:
Seems one test is failing: 
```
# fail because value of one param mandates presence of other param required
        req_if = (('foo', 'hello', ('bam')),)
        args = json.dumps(dict(ANSIBLE_MODULE_ARGS={'foobar': [{"foo": "hello"}]}))
        with swap_stdin_and_argv(stdin_data=args):
            basic._ANSIBLE_ARGS = None
            self.assertRaises(
                SystemExit,
                basic.AnsibleModule,
                argument_spec=arg_spec,
                required_if=req_if,
                no_log=True,
                check_invalid_arguments=False,
                add_file_common_args=True,
>               supports_check_mode=True
            )
E           AssertionError: SystemExit not raised by AnsibleModule
```

However, that test, in my opinion, is wrong: The test starts with a global req_if, but then uses a subspec.

Bunch of other tests show the same issue, it seems. I'll look if i can fix the test tomorrow.

ganeshrn reviewed (commented):
I think a similar check is applicable for `mutually_exclusive` at L1954.
Other changes look good me.

gundalow commented:
Unit tests need updating to defend this functionality https://github.com/ansible/ansible/blob/devel/test/units/module_utils/test_basic.py

Most was covered in https://github.com/ansible/ansible/commit/97a34cf008501088ae012f24e4712e2bc6dd6d4c#diff-26ddd7713399e234f0d8337d0cc19a63

pdellaert committed:
Also fixing mutually exclusive check

pdellaert committed:
Updating subspec checks
These changes take into account a spec with all features enabled and do
the following tests for subspecs:
1. Test proper specs
2. Test Alias
3. Test missing required param
4. Test mutually exclusive params
5. Test required if params
6. Test required one of params
7. Test required together params
8. Test required if params with a default value
9. Test basis subspec params
10. Test invalid subsec params

ganeshrn commented:
@pdellaert The test needs to be updated as per this fix. Initially, it was written to use global conditional check for subspec.
With this fix the sub arg spec in unit test need to be modified as 
```
req_to = (('bam', 'baz'),)
- arg_spec = dict(foobar=dict(type='list', elements='dict', options=options_spec))
+ arg_spec = dict(foobar=dict(type='list', elements='dict', options=options_spec), req_to=req_to)
```

pdellaert commented:
@ganeshrn @gundalow @bcoca 
I've updated the tests and the `mutually_exclusive`, in my commit message on the tests, you can see the tests that i'm executing and the logic behind those. overview:
```
1. Test proper specs
2. Test Alias
3. Test missing required param
4. Test mutually exclusive params
5. Test required if params
6. Test required one of params
7. Test required together params
8. Test required if params with a default value
9. Test basis subspec params
10. Test invalid subsec params
```

pdellaert commented:
The one test that fails is where yum fails to download a package (404 error), don't think that should impact this
https://app.shippable.com/github/ansible/ansible/runs/31746/14/tests

ganeshrn commented:
Re-triggering build by closing and opening PR.

pdellaert commented:
same: https://app.shippable.com/github/ansible/ansible/runs/31755/15/tests , yum keeps getting 404 errors on the mirrors for `TASK [yum : install sos]` (only on Fedora25)

ganeshrn commented:
@nitzmahone @bcoca @gundalow Merging this PR

@pdellaert Thank you!!!
